File: api/src/main/java/com/api/app/common/response/ApiResponse.java
package com.api.app.common.response;

import com.api.app.common.exception.ApiError;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Getter;

import java.time.LocalDateTime;

/**
 * @author system
 * @version 1.0
 * @since 2025-10-28
 */
@Getter
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final LocalDateTime timestamp;
    private final String code;
    private final String message;
    private final T data;

    public ApiResponse(String code, String message, T data) {
        this.timestamp = LocalDateTime.now();
        this.code = code;
        this.message = message;
        this.data = data;
    }

    /**
     * 성공 응답 (데이터 포함)
     *
     * @param data 응답 데이터
     * @param <T>  데이터 타입
     * @return ApiResponse
     */
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(ApiError.SUCCESS.getCode(), ApiError.SUCCESS.getMessage(), data);
    }

    /**
     * 성공 응답 (데이터 없음)
     *
     * @param <T> 데이터 타입
     * @return ApiResponse
     */
    public static <T> ApiResponse<T> success() {
        return new ApiResponse<>(ApiError.SUCCESS.getCode(), ApiError.SUCCESS.getMessage(), null);
    }

    /**
     * 실패 응답
     *
     * @param apiError 에러 정보
     * @param <T>      데이터 타입
     * @return ApiResponse
     */
    public static <T> ApiResponse<T> error(ApiError apiError) {
        return new ApiResponse<>(apiError.getCode(), apiError.getMessage(), null);
    }

    /**
     * 실패 응답 (커스텀 메시지)
     *
     * @param apiError      에러 정보
     * @param customMessage 커스텀 메시지
     * @param <T>           데이터 타입
     * @return ApiResponse
     */
    public static <T> ApiResponse<T> error(ApiError apiError, String customMessage) {
        return new ApiResponse<>(apiError.getCode(), customMessage, null);
    }
}

File: api/src/main/kotlin/com/api/app/common/response/ApiResponse.kt
package com.api.app.common.response

import com.api.app.common.exception.ApiError
import com.fasterxml.jackson.annotation.JsonInclude
import java.time.LocalDateTime

@JsonInclude(JsonInclude.Include.NON_NULL)
data class ApiResponse<T>(
    val timestamp: LocalDateTime = LocalDateTime.now(),
    val code: String,
    val message: String,
    val data: T? = null
) {
    companion object {
        fun <T> success(data: T): ApiResponse<T> =
            ApiResponse(code = ApiError.SUCCESS.code, message = ApiError.SUCCESS.message, data = data)

        fun <T> success(): ApiResponse<T> =
            ApiResponse(code = ApiError.SUCCESS.code, message = ApiError.SUCCESS.message)

        fun <T> error(apiError: ApiError): ApiResponse<T> =
            ApiResponse(code = apiError.code, message = apiError.message)

        fun <T> error(apiError: ApiError, customMessage: String): ApiResponse<T> =
            ApiResponse(code = apiError.code, message = customMessage)
    }
}
